sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/AvroGenericRecordToStorageApiProto.java

```diff
@@ -0,0 +1,382 @@
+                .addFields(keyFieldSchema)
+                .addFields(valueFieldSchema)
+                .setMode(TableFieldSchema.Mode.REPEATED);
+        break;
+      case UNION:
+        elementType = TypeWithNullability.create(schema).getType();
+        if (elementType == null) {
+          throw new RuntimeException("Unexpected null element type!");
+        }
+        // check to see if more than one non-null type is defined in the union
+        Preconditions.checkState(
+            elementType.getType() != Schema.Type.UNION,
+            "Multiple non-null union types are not supported.");
+        TableFieldSchema unionFieldSchema =
+            fieldDescriptorFromAvroField(
+                new Schema.Field(field.name(), elementType, field.doc(), field.defaultVal()));
+        builder =
+            builder
+                .setType(unionFieldSchema.getType())
+                .addAllFields(unionFieldSchema.getFieldsList());
+        break;
+      default:
+        elementType = TypeWithNullability.create(schema).getType();
+        @Nullable
+        TableFieldSchema.Type primitiveType =
+            Optional.ofNullable(LogicalTypes.fromSchema(elementType))
+                .map(logicalType -> LOGICAL_TYPES.get(logicalType.getName()))
+                .orElse(PRIMITIVE_TYPES.get(elementType.getType()));
+        if (primitiveType == null) {
+          throw new RuntimeException("Unsupported type " + elementType.getType());
+        }
+        // a scalar will be required by default, if defined as part of union then
+        // caller will set nullability requirements
+        builder = builder.setType(primitiveType);
+    }
+    if (builder.getMode() != TableFieldSchema.Mode.REPEATED) {
+      if (TypeWithNullability.create(schema).isNullable()) {
+        builder = builder.setMode(TableFieldSchema.Mode.NULLABLE);
+      } else {
+        builder = builder.setMode(TableFieldSchema.Mode.REQUIRED);
+      }
+    }
+    if (field.doc() != null) {
+      builder = builder.setDescription(field.doc());
+    }
+    return builder.build();
+  }
+
+  @Nullable
+  private static Object messageValueFromGenericRecordValue(
+      FieldDescriptor fieldDescriptor, Schema.Field avroField, String name, GenericRecord record) {
+    @Nullable Object value = record.get(name);
+    if (value == null) {
+      if (fieldDescriptor.isOptional()) {
+        return null;
+      } else {
+        throw new IllegalArgumentException(
+            "Received null value for non-nullable field " + fieldDescriptor.getName());
+      }
+    }
+    return toProtoValue(fieldDescriptor, avroField.schema(), value);
+  }
+
+  private static Object toProtoValue(
+      FieldDescriptor fieldDescriptor, Schema avroSchema, Object value) {
+    switch (avroSchema.getType()) {
+      case RECORD:
+        return messageFromGenericRecord(fieldDescriptor.getMessageType(), (GenericRecord) value);
+      case ARRAY:
+        List<Object> list = (List<Object>) value;
```

Comment:
Are we sure it's always a List (and not just an Iterable)?

Reply:
Changed it to Iterable.